File: java/hw/hw3/src/trefoil2/Interpreter.java
package trefoil2;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.val;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * Interprets expressions and bindings in the context of a dynamic environment
 * according to the semantics of Trefoil v2.
 */
public class Interpreter {

    private static DynamicEnvironment.Entry entry;
    private static DynamicEnvironment.Entry entry1;

    /**
     * Evaluates e in the given environment. Returns the resulting value.
     *
     * Throws TrefoilError.RuntimeError when the Trefoil programmer makes a mistake.
     */
    public static Expression interpretExpression(Expression e, DynamicEnvironment environment) {
        if (e instanceof Expression.IntegerLiteral) {
            return e;
        } else if (e instanceof Expression.VariableReference) {
            Expression.VariableReference var = (Expression.VariableReference) e;
            return environment.getVariable(var.getVarname());
        } else if (e instanceof Expression.Plus) {
            Expression.Plus p = (Expression.Plus) e;
            Expression v1 = interpretExpression(p.getLeft(), environment);
            Expression v2 = interpretExpression(p.getRight(), environment);
            // TODO: the following return statement is wrong because it does not correctly check
            //       for run-time type errors. fix it by checking that both children evaluated to
            //       IntegerLiterals and if not throwing TrefoilError.RuntimeError.

            if (v1 instanceof Expression.IntegerLiteral && v2 instanceof Expression.IntegerLiteral) {
                return new Expression.IntegerLiteral(
                        ((Expression.IntegerLiteral) v1).getData() +
                                ((Expression.IntegerLiteral) v2).getData()
                );
            }
                throw new Trefoil2.TrefoilError.RuntimeError("There is one or more none integer values");


        } else if (e instanceof Expression.BooleanLiteral) {

            // TODO: implement semantics for new AST nodes here, following the examples above
            // TODO: be sure to check for run-time type errors and throw TrefoilError.RuntimeError.

            if (e.toString() != "true" && e.toString() != "false") {
                throw new Trefoil2.TrefoilError.RuntimeError("This is not a boolean value");
            } else {
                return e;
            }


        } else if (e instanceof Expression.Times) {


            Expression.Times p = (Expression.Times) e;
            Expression v1 = interpretExpression(p.getLeft(), environment);
            Expression v2 = interpretExpression(p.getRight(), environment);

            if (v1 instanceof Expression.IntegerLiteral && v2 instanceof Expression.IntegerLiteral) {
                return new Expression.IntegerLiteral(
                        ((Expression.IntegerLiteral) v1).getData() *
                                ((Expression.IntegerLiteral) v2).getData()
                );
            }

            throw new Trefoil2.TrefoilError.RuntimeError("There is one or more none integer values");


        } else if (e instanceof Expression.Minus) {

            Expression.Minus p = (Expression.Minus) e;
            Expression v1 = interpretExpression(p.getLeft(), environment);
            Expression v2 = interpretExpression(p.getRight(), environment);


            if (v1 instanceof Expression.IntegerLiteral && v2 instanceof Expression.IntegerLiteral) {
                return new Expression.IntegerLiteral(
                        ((Expression.IntegerLiteral) v1).getData() -
                                ((Expression.IntegerLiteral) v2).getData()
                );
            }
            throw new Trefoil2.TrefoilError.RuntimeError("There is one or more none integer values");

        } else if (e instanceof Expression.Equals) {

            Expression.Equals p = (Expression.Equals) e;
            Expression e1 = interpretExpression(p.getLeft(), environment);
            Expression e2 = interpretExpression(p.getRight(), environment);


            if (e1 instanceof Expression.IntegerLiteral && e2 instanceof Expression.IntegerLiteral) {

                boolean test = e1.equals(e2);

                return new Expression.BooleanLiteral(test);
            }
            throw new Trefoil2.TrefoilError.RuntimeError("There is one or more none integer values");

        } else if (e instanceof Expression.Cons) {


            Expression.Cons temp = (Expression.Cons) e;

            Expression e1 = interpretExpression(temp.getValOne(), environment);
            Expression e2 = interpretExpression(temp.getValTwo(), environment);

                return Expression.cons(e1, e2);

        } else if (e instanceof Expression.Nil) {

            return new Expression.Nil();

        } else if (e instanceof Expression.NilQuestion) {

            Expression.NilQuestion temp = (Expression.NilQuestion) e;
            Expression temp2 = interpretExpression(temp.getBody(), environment);

            if (temp2 instanceof Expression.Nil) {
                return new Expression.BooleanLiteral(true);
            }
            return new Expression.BooleanLiteral(false);

        } else if (e instanceof Expression.ConsQuestion) {

            Expression.ConsQuestion temp = (Expression.ConsQuestion) e;
            Expression temp2 = interpretExpression(temp.getBody(), environment);

            if (temp2 instanceof Expression.Cons) {
                return new Expression.BooleanLiteral(true);
            }
            return new Expression.BooleanLiteral(false);

        } else if (e instanceof Expression.Car) {
            Expression.Car temp = (Expression.Car) e;
            Expression temp2 = interpretExpression(temp.getBody(), environment);


         if (temp2 instanceof Expression.Cons) {
                Expression.Cons temp3 = (Expression.Cons) temp2;
                return interpretExpression(temp3.getValOne(), environment);
         }

            throw new Trefoil2.TrefoilError.RuntimeError("Requires a Cons Value");

        } else if (e instanceof Expression.Cdr) {
            Expression.Cdr temp = (Expression.Cdr) e;
            Expression temp2 = interpretExpression(temp.getBody(), environment);

            if (temp2 instanceof Expression.Cons) {
                Expression.Cons temp3 = (Expression.Cons) temp2;
                return interpretExpression(temp3.getValTwo());
            }

           throw new Trefoil2.TrefoilError.RuntimeError("Requires a Cons Value");

        } else if (e instanceof Expression.If) {


            Expression.If p = (Expression.If) e;
            Expression e1 = interpretExpression(p.getCondition(), environment);

            if (e1 instanceof Expression.BooleanLiteral) {

               String value = ((Expression.BooleanLiteral) e1).toString();
                        if(value.equals("false")) {

                        Expression e3 = interpretExpression(p.getFalseCondition(), environment);
                           return e3;
                        }
            }


            Expression e2 = interpretExpression(p.getTrueCondition(), environment);

            return e2;

        } else if (e instanceof Expression.Let) {

            Expression.Let p = (Expression.Let) e;
            String test = p.getVarName();
            Expression test2 = p.getVarDef();

            DynamicEnvironment newEnv = environment.extendVariable(p.getVarName(), interpretExpression(p.getVarDef(), environment));
            return interpretExpression(p.getBody(), newEnv);

        } else if (e instanceof Expression.FunctionCall) {

            Expression.FunctionCall p = (Expression.FunctionCall) e;
            String funName = p.getFunName();
            List<Expression> body = p.getBody();


            DynamicEnvironment callenv = environment;
            DynamicEnvironment defenv;
            Binding.FunctionBinding binding;

            if(callenv.containsFunction(funName)) {
                DynamicEnvironment.Entry.FunctionEntry ultimateFunction = callenv.getFunction(funName);
                binding = ultimateFunction.getFunctionBinding();
                defenv = ultimateFunction.getDefiningEnvironment();
            } else {
                throw new Trefoil2.TrefoilError.RuntimeError("This function does not exist");
            }

            List<String> args = binding.getArgnames();

            if(body.size() != args.size()) {
                throw new Trefoil2.TrefoilError.RuntimeError("Incorrect number of parameters entered");
            }

            List<Expression> vals = new LinkedList<>();
            for(int i = 0; i < body.size(); i++) {
                vals.add(interpretExpression(body.get(i), callenv));
              }

            DynamicEnvironment test = environment.extendVariables(args, body);



           return interpretExpression(binding.getBody(), defenv.extendVariables(args, vals));

            /* Consider a function call expression (f args), where f stands for any
                function name, and args stands for any sequence (possibly empty) of
                expressions to be passed as arguments to f. Call the current dynamic
                environment callenv. The semantics is to first lookup f in callenv. If
                f is not mapped to anything, or maps to anything besides a function entry,
                signal an error. From the function entry mapped to by f, let (define (f params) body)
                be f's function binding and defenv its defining
                environment.  Next, evaluate each element of args in left-to-right order in
                callenv. This results in a sequence of values. Call that sequence vals.
                Then evaluate body in a new dynamic environment constructed by extending
                defenv with mappings param -> val for each param and corresponding value
                val in the lists params and values. Return the result of evaluating
                body.
                This is the hardest part of the whole assignment. Do not be surprised if you
                have to read that paragraph 5 or 10 times.*/
        } else {
            // Otherwise it's an expression AST node we don't recognize. Tell the interpreter implementor.
            throw new Trefoil2.InternalInterpreterError("\"impossible\" expression AST node " + e.getClass());
        }
    }

    /**
     * Executes the binding in the given environment, returning the new environment.
     *
     * The environment passed in as an argument is *not* mutated. Instead, it is copied
     * and any modifications are made on the copy and returned.
     *
     * Throws TrefoilError.RuntimeError when the Trefoil programmer makes a mistake.
     */

    public static DynamicEnvironment interpretBinding(Binding b, DynamicEnvironment environment) {
        if (b instanceof Binding.VariableBinding) {
            Binding.VariableBinding vb = (Binding.VariableBinding) b;
            Expression value = interpretExpression(vb.getVardef(), environment);
            System.out.println(vb.getVarname() + " = " + value);
            return environment.extendVariable(vb.getVarname(), value);
        } else if (b instanceof Binding.TopLevelExpression) {
            Binding.TopLevelExpression tle = (Binding.TopLevelExpression) b;
            System.out.println(interpretExpression(tle.getExpression(), environment));
            return environment;
        } else if (b instanceof Binding.FunctionBinding) {
            Binding.FunctionBinding fb = (Binding.FunctionBinding) b;
            DynamicEnvironment newEnvironment = environment.extendFunction(fb.getFunname(), fb);
            System.out.println(fb.getFunname() + " is defined");
            return newEnvironment;
        } else if (b instanceof Binding.TestBinding) {
            // TODO: implement the TestBinding here
            Binding.TestBinding v = (Binding.TestBinding) b;
            Expression value = interpretExpression(v.getExpression(), environment);

            if(!(value instanceof Expression.BooleanLiteral) || ((Expression.BooleanLiteral) value).toString().equals("false")) {
                throw new Trefoil2.TrefoilError.RuntimeError("No such value exists");
            }

            return environment;

        }



        // Otherwise it's a binding AST node we don't recognize. Tell the interpreter implementor.
        throw new Trefoil2.InternalInterpreterError("\"impossible\" binding AST node " + b.getClass());
    }


    // Convenience methods for interpreting in the empty environment.
    // Used for testing.
    public static Expression interpretExpression(Expression e) {
        return interpretExpression(e, new DynamicEnvironment());
    }
    public static DynamicEnvironment interpretBinding(Binding b) {
        return interpretBinding(b, DynamicEnvironment.empty());
    }


    /**
     * Represents the dynamic environment, which is a mapping from strings to "entries".
     * In the starter code, the string always represents a variable name and an entry is always a VariableEntry.
     * You will extend it to also support function names and FunctionEntries.
     */
    @Data
    public static class DynamicEnvironment {

        private Entry entry;
        private Entry entry2;

        public static abstract class Entry {
            @EqualsAndHashCode(callSuper = false)
            @Data
            public static class VariableEntry extends Entry {
                private final Expression value;
            }

            @EqualsAndHashCode(callSuper = false)
            @Data
            public static class FunctionEntry extends Entry {
                private final Binding.FunctionBinding functionBinding;

                @ToString.Exclude
                private final DynamicEnvironment definingEnvironment;
            }

            // Convenience factory methods

            public static Entry variable(Expression value) {
                return new VariableEntry(value);
            }
            public static Entry function(Binding.FunctionBinding functionBinding, DynamicEnvironment definingEnvironment) {
                return new FunctionEntry(functionBinding, definingEnvironment);
            }
        }

        // The backing map of this dynamic environment.
        private final Map<String, Entry> map;

        public DynamicEnvironment() {
            this.map = new HashMap<>();
        }

        public DynamicEnvironment(DynamicEnvironment other) {
            this.map = new HashMap<>(other.getMap());
        }

        private boolean containsVariable(String varname) {
            return map.containsKey(varname) && map.get(varname) instanceof Entry.VariableEntry;
        }

        public Expression getVariable(String varname) {
            // TODO: convert this assert to instead throw a TrefoilError.RuntimeError if the variable is not bound


            if(!containsVariable(varname)) {
                 throw new Trefoil2.TrefoilError.RuntimeError("There is no variable called " + varname + " bound");
            }



            // TODO: lookup the variable in the map and return the corresponding value
            // Hint: first, read the code for containsVariable().
            // Hint: you will likely need the value field from Entry.VariableEntry
            return ((Entry.VariableEntry) map.get(varname)).getValue();
        }

        public void putVariable(String varname, Expression value) {
            // TODO: bind the variable in the backing map
            // Hint: map.put
            // Hint: either call new Entry.VariableEntry or the factory Entry.variable
            map.put(varname, Entry.variable(value));
        }

        /**
         * Returns a *new* DynamicEnvironment extended by the binding varname -> value.
         *
         * Does not change this! Creates a copy.
         */
        public DynamicEnvironment extendVariable(String varname, Expression value) {
            DynamicEnvironment newEnv = new DynamicEnvironment(this);  // create a copy
            newEnv.putVariable(varname, value);  // mutate the copy
            return newEnv;  // return the mutated copy (this remains unchanged!)
        }

        /**
         * Returns a *new* Dynamic environment extended by the given mappings.
         *
         * Does not change this! Creates a copy.
         *
         * varnames and values must have the same length
         *
         * @param varnames variable names to bind
         * @param values values to bind the variables to
         */
        public DynamicEnvironment extendVariables(List<String> varnames, List<Expression> values) {
            DynamicEnvironment newEnv = new DynamicEnvironment(this);
            assert varnames.size() == values.size();
            for (int i = 0; i < varnames.size(); i++) {
                newEnv.putVariable(varnames.get(i), values.get(i));
            }
            return newEnv;
        }

        private boolean containsFunction(String funname) {
            return map.containsKey(funname) && map.get(funname) instanceof Entry.FunctionEntry;
        }

        public Entry.FunctionEntry getFunction(String funname) {
            // TODO: convert this assert to instead throw a TrefoilError.RuntimeError if the function is not bound
            if(!containsFunction(funname)) {
             throw new Trefoil2.TrefoilError.RuntimeError("There is no function named " + funname + "bound");
            }



            // TODO: lookup the function in the map and return the corresponding function binding
            // Hint: first, read the code for containsFunction().

            Entry.FunctionEntry temp = (Entry.FunctionEntry) map.get(funname);
            return temp;
        }

        public void putFunction(String funname, Binding.FunctionBinding functionBinding) {
            // TODO: bind the function in the backing map
            // Be careful to set up recursion correctly!
            // Hint: Pass definingEnvironment=this to the Entry.function factory, and then call map.put.
            //       That way, by the time Trefoil calls the function, everything points to
            //       the right place. Tricky!

            map.put(funname, new Entry.FunctionEntry(functionBinding, this));


        }

        public DynamicEnvironment extendFunction(String funname, Binding.FunctionBinding functionBinding) {
            DynamicEnvironment newEnv = new DynamicEnvironment(this);  // create a copy of this
            newEnv.putFunction(funname, functionBinding);  // mutate the copy
            return newEnv;  // return the copy
        }

        // Convenience factory methods

        public static DynamicEnvironment empty() {
            return new DynamicEnvironment();
        }

        public static DynamicEnvironment singleton(String varname, Expression value) {
            return empty().extendVariable(varname, value);
        }
    }
}

File: ocaml/hw5/src/interpreter.ml
open Ast
open Errors

type entry =
  | VariableEntry of expr
  | FunctionEntry of function_binding * dynamic_env
  | StructEntry of struct_binding
[@@deriving show]
and dynamic_env = (string * entry) list
let entry_of_string = show_entry

let rec lookup (dynenv, name) =
  match dynenv with
  | [] -> None
  | (x, entry) :: dynenv ->
     if x = name
     then Some entry
     else lookup (dynenv, name)

(* ignore this until working on part 2 *)
let rec interpret_pattern (pattern, value): (string * entry) list option =
  match pattern, value with
  
  | IntLitPattern n, IntLit num  -> begin
    if n = num
      then Some []
    else None

    end
  | BoolLitPattern b, BoolLit boo -> begin
    if b = boo
      then Some []
     else None


      end
  | NilPattern, Nil ->  Some []
  | SymbolPattern s, Symbol s2 -> begin
    
      if s = s2
        then Some []
      else None  
    end  
    

  | VariablePattern v, x -> Some [(v, (VariableEntry x))]
  | StructPattern (namePat, patterns), StructConstructor(nameCon, expressions) -> begin
    if(namePat <> nameCon)
      then None
  else if (List.length(patterns) <> List.length(expressions))
    then None
  else
    let rec looper (pi, vi, acc) =
      match pi with
      | [] -> List.rev(acc)
      | x :: xs -> 
        match vi with
        | [] -> List.rev(acc)
        | y :: ys ->
          match interpret_pattern(x, y) with
          | Some s -> looper (xs, ys, s @ acc)
          | None -> looper(xs, ys, acc)
 in

 let newEnv =  looper(patterns, expressions, []) in

 if(List.length(newEnv) < List.length(patterns))
  then None
else

    Some(newEnv)

  end
  | WildcardPattern, _ -> Some []
  | ConsPattern (p1, p2), Cons (v1, v2) -> begin
     match interpret_pattern (p1, v1), interpret_pattern (p2, v2) with
     | Some l1, Some l2 -> Some (l1 @ l2)
     | _ -> None
    end
  
  (* TODO: add cases for other kinds of patterns here *)

  | _ -> None

let rec interpret_expression (dynenv, e) =
  (* helper function to interpret a list of expressions into a list of values *)
  let rec interpret_list (dynenv, es) =
    match es with
    | [] -> []
    | e :: es -> interpret_expression (dynenv, e) :: interpret_list (dynenv, es)
  in
  match e with
  | IntLit _ | BoolLit _ | Nil | StructConstructor _ -> e
  | Variable x -> begin
      match lookup (dynenv, x) with
      | None -> raise (RuntimeError ("Unbound variable " ^ x))
      | Some (VariableEntry value) -> value
      | Some e -> raise (RuntimeError ("Expected name " ^ x ^ " to refer to a variable, but got something else: " ^ entry_of_string e))
    end
  | Plus (e1, e2) -> begin
      match interpret_expression (dynenv, e1), interpret_expression (dynenv, e2) with
      | IntLit n1, IntLit n2 -> IntLit (n1 + n2)
      | IntLit _, v2 -> raise (RuntimeError ("Plus applied to non-integer " ^ string_of_expr v2))
      | v1, _ -> raise (RuntimeError ("Plus applied to non-integer " ^ string_of_expr v1))
    end
  | Minus (e1, e2) -> begin
      match interpret_expression (dynenv, e1), interpret_expression (dynenv, e2) with
      | IntLit n1, IntLit n2 -> IntLit (n1 - n2)
      | IntLit _, v2 -> raise (RuntimeError ("Minus applied to non-integer " ^ string_of_expr v2))
      | v1, _ -> raise (RuntimeError ("Minus applied to non-integer " ^ string_of_expr v1))
    end
  | Times (e1, e2) -> begin
      match interpret_expression (dynenv, e1), interpret_expression (dynenv, e2) with
      | IntLit n1, IntLit n2 -> IntLit (n1 * n2)
      | IntLit _, v2 -> raise (RuntimeError ("Times applied to non-integer " ^ string_of_expr v2))
      | v1, _ -> raise (RuntimeError ("Times applied to non-integer " ^ string_of_expr v1))
    end

  | Equals (e1, e2) -> begin
      match interpret_expression (dynenv, e1), interpret_expression (dynenv, e2) with
        | n1, n2 -> BoolLit (n1 = n2)
      end
  | If (e1, e2, e3) -> begin
      match interpret_expression (dynenv, e1) with
      | BoolLit false -> interpret_expression (dynenv, e3)
      | _ -> interpret_expression (dynenv, e2)
    end
  (* TODO: add case for let expressions here *)
  | Let (e1, e2, e3) -> begin
      match (e1, e2, e3) with
      | ((varname : string), (vardef : expr), (body : expr)) -> 
        (let newenv =  (varname, (VariableEntry vardef )) :: dynenv in
         interpret_expression(newenv, body))
      | _ -> raise (RuntimeError ("Let expressions require a string and two expressions"))
  end
  | Cons (e1, e2) ->
     let v1 = interpret_expression (dynenv, e1) in
     let v2 = interpret_expression (dynenv, e2) in
     Cons (v1, v2)
  | IsNil e -> begin
     match interpret_expression (dynenv, e) with
     | Nil -> BoolLit true
     | _ -> BoolLit false
    end
  | IsCons e -> begin
     match interpret_expression (dynenv, e) with
     | Cons _ -> BoolLit true
     | _ -> BoolLit false
    end
  | Car e -> begin
     match interpret_expression (dynenv, e) with
     | Cons (v1, _)  -> v1
     | v -> raise (RuntimeError("car applied to non-cons " ^ string_of_expr v))
    end
  | Cdr e -> begin
     match interpret_expression (dynenv, e) with
     | Cons (_, v2)  -> v2
     | v -> raise (RuntimeError("car applied to non-cons " ^ string_of_expr v))
    end

  | StructPredicate (s, e) -> begin

    match interpret_expression (dynenv, e) with
    | StructConstructor(s', vs) -> BoolLit (s = s')
    | _ -> BoolLit false

    (*Consider a struct predicate expression StructPredicate(s, e) where s
stands for any string and e stands for any expression AST. Evaluate e in
the current dynamic environment to a value v. If v is of the form
StructConstructor(s', vs) and s and s' are equal as strings, return
true. In all other cases, return false.*)

    end

  | StructAccess ((s : string), (i : int), (e : expr)) -> begin

    match interpret_expression (dynenv, e) with
    | StructConstructor(s', vs) ->
        let rec looper(count, lister) : expr =
          match lister with
          | [] -> raise (RuntimeError("shouldn't happen"))
          | temp::temps ->
            if (count = 0)
            then temp
            else  looper(count-1, temps)
        in
        if ((s = s') && (i < (List.length vs)))
          then 
            looper(i, vs)
          else 
            raise (RuntimeError("Either the strings don't match or index is longer than the length of the list."))

    | _ -> raise (RuntimeError("Not a StructConstructor"))

    (*Consider a struct access expression StructAccess(s, i, e) where s stands
for any string, i stands for any (OCaml) integer, and e stands for any
expression AST. Evaluate e in the current dynamic environment to a value
v. If v is of the form StructConstructor(s', vs) and s and s' are
equal as strings and i is less than the length of vs, then return the
ith element of vs. In all other cases, signal an error.*)

  end

  | Call (fun_name, arg_exprs) -> begin
      let callenv = dynenv in
      match lookup (callenv, fun_name) with
      | None -> raise (RuntimeError ("Unbound function " ^ fun_name))
      | Some ((FunctionEntry (fb, defenv)) as entry) ->
         (* This provided line handles recursion differently than (but
            equivalent to) HW3! When you build the environment to evaluate the
            function body, start from this defenv. *)
            let argnames = fb.param_names in
            let body = fb.body in
            if (List.length(argnames) != List.length(arg_exprs))
          then 
            raise (RuntimeError("Incorrect number of arguments"))
         else
            let defenv = (fun_name, entry) :: defenv in
           
            let rec arglooper (args, acc) =
              (match args with
              | [] -> List.rev(acc)
              | x::xs -> arglooper(xs, interpret_expression(dynenv, x) :: acc)) in 
            let vals = arglooper(arg_exprs, []) in
            let rec bindinglooper (names, values, acc) =
              (match names with
                | [] -> List.rev(acc)
                | x :: xs -> (match values with
                                | [] -> List.rev(acc)
                                | y :: ys -> bindinglooper(xs, ys, (x, VariableEntry (y)) :: acc)  
                            )
              )
          in 
           let newenv = defenv @ bindinglooper(argnames, vals, [])  in
          interpret_expression(newenv, body)
        
      | Some (StructEntry sb) ->
         (* ignore this until part 2 *)

         if((List.length sb.field_names) = (List.length arg_exprs))
          then          
            
            let rec arglooper (args, acc) =
            (match args with
            | [] -> List.rev(acc)
            | x::xs -> arglooper(xs, interpret_expression(dynenv, x) :: acc)) in 
            let vals = arglooper(arg_exprs, []) in

            StructConstructor(fun_name, vals)

         else
          raise (RuntimeError("Incorrect number of arguments"))          


        (*Otherwise, if f maps to a struct entry (whose name
        is the same as f), evaluate each element of args in left-to-right order in
        callenv. Call the resulting sequence of values vals. Return
        StructConstructor(f, vals).*)

    | Some e -> raise (RuntimeError ("Expected name " ^ fun_name ^ " to refer to a function or struct, but got something else: " ^ entry_of_string e))
    end
  | Cond clauses ->
     let rec loop clauses =
       match clauses with
       | [] -> raise (RuntimeError("cond failure: no clauses left"))
       | (predicate, body) :: clauses ->
          if(BoolLit true = interpret_expression(dynenv, predicate))
            then interpret_expression(dynenv, body)
        else
          loop(clauses)
      
      in
     loop clauses
  | Symbol _ -> e

  (* TODO: add cases for the other "internal" expressions here *)
  (* TODO: add case for match expressions here *)
  | Match (e, clauses) -> begin

    let v = interpret_expression(dynenv, e) in

    let rec loop arguments =
      match arguments with
      | [] -> raise (RuntimeError("match failure: no clauses left"))
      | ((pi : pattern), bi) :: clausers ->

          match interpret_pattern(pi, v) with
          | Some args ->
            let newenv = args @ dynenv in
            interpret_expression(newenv, bi)
          | _ -> loop(clausers)
     
     in
     loop(clauses)
  end

  

let interpret_binding (dynenv, b) =
  match b with
  | VarBinding (x, e) ->
     let v = interpret_expression (dynenv, e) in
     Printf.printf "%s = %s\n%!" x (string_of_expr v);
     (x, VariableEntry v) :: dynenv
  | TopLevelExpr e ->
     let v = interpret_expression (dynenv, e) in
     print_endline (string_of_expr v);
     dynenv
  | FunctionBinding fb ->
     Printf.printf "%s is defined\n%!" fb.name;
     (fb.name, FunctionEntry (fb, dynenv)) :: dynenv
  (* TODO: implement test bindings here *)
  | TestBinding tb ->
    let temp = interpret_expression(dynenv, tb) in
      (match temp with
      | BoolLit true -> dynenv
      | _ -> raise (RuntimeError("No such value exists"))
      
      )
  | StructBinding sb ->
     (* TODO: uncomment the comment on the next line and replace the "..." with
        a mapping for the structs name to a StructEntry containing sb. *)
     let dynenv =
      
      (sb.name, StructEntry(sb)) :: dynenv in

 


     (* TODO: create struct predicate function here *)

     let func =  {name = (sb.name ^ "?"); param_names = [("x" : string)];
      body = Ast.StructPredicate(sb.name, (Ast.Variable("x")))} in
     let dynenv = ((sb.name ^ "?"), FunctionEntry(func, dynenv)) :: dynenv in


      (*Back in the StructBinding case for interpret_binding, generate a
function binding for the struct's predicate function.

Hint: The function's definition is described in LANGUAGE.md "as if by the
binding ..." with a binding that is not valid user-facing syntax because it
uses StructPredicate. So, you cannot use the parser to generate this
binding, and you must instead construct it manually using the AST
constructors for function bindings, StructPredicate, and variables.*)


     (* TODO: uncomment this when ready to do accessor functions *)
       
     let fun_entry_for_accessor (idx, field_name): string * entry =
      (sb.name ^ "-" ^ field_name, 
      FunctionEntry({
        name = (sb.name ^ "-" ^ field_name);
        param_names = [("x" : string)];
        body = Ast.StructAccess(sb.name, idx, (Ast.Variable("x")))},
        dynenv)
      ) 
     in
 

     let rec fun_entry_accessor_loop (idx, field_names) =
       match field_names with
       | [] -> []
       | f :: field_names -> fun_entry_for_accessor (idx, f) :: fun_entry_accessor_loop (idx+1, field_names)
     in
     
     let dynenv = fun_entry_accessor_loop (0, sb.field_names) @ dynenv in

     
(*(define (s? x) StructPredicate(s, x))*)     
(*(define (s-f x) StructAccess(s, i, x)) 
for each field f in fs, s-f maps to a function defined as if by the
binding (define (s-f x) StructAccess(s, i, x)) where i is the index of
the field f in the sequence fs.*)

     dynenv

(* the semantics of a whole program (sequence of bindings) *)
let rec interpret_bindings (dynenv, bs) =
  match bs with
  | [] -> dynenv
  | b :: bs ->
     interpret_bindings (interpret_binding (dynenv, b), bs)

(* starting from dynenv, first interpret the list of bindings in order. then, in
   the resulting dynamic environment, interpret the expression and return its
   value *)
let interpret_expression_after_bindings (dynenv, bindings, expr) =
  interpret_expression (interpret_bindings (dynenv, bindings), expr)
